testng-core-api/src/main/java/org/testng/IParameterInfo.java

```diff
@@ -0,0 +1,36 @@
+package org.testng;
+
+/** Represents the ability to retrieve the parameters associated with a factory method. */
+public interface IParameterInfo {
+
+  /** @return - The actual instance associated with a factory method */
+  Object getInstance();
+
+  /**
+   * @return - The actual index of instance associated with a factory method. This index has a 1:1
+   *     correspondence with what were specified via the <code>indices</code> attribute of the
+   *     <code>@Factory</code> annotation. For e.g., lets say you specified the indices to the "1"
+   *     and your factory returned 4 instances, then the instance on which this method is invoked
+   *     would have the value as "1".
+   */
+  int getIndex();
+
+  /**
+   * @return - returns an index which indicates the running position in the array of test class
+   *     instances that were produced by a <code>@Factory</code> annotated constructor or static
+   *     method. For e.g., lets say your <code>@Factory</code> method returned 4 instances, then
+   *     each of the invocations to this method would return a value from <code>0</code> to <code>3
+   *     </code>
+   */
+  int currentIndex();
+
+  /** @return - The parameters associated with the factory method as an array. */
+  Object[] getParameters();
```

Comment:
As it is the parameter of the factory method, it should not be located here.

Reply:
So you want to have this moved over to ITestNGMethod interface ?

Reply:
`FactoryMethod` is an implementation of `ITestNGMethod`. So I wouldn't be able to put this inside `FactoryMethod` because then this will NOT be exposed to our users and so they wont be able to access it. We currently expose the `IParameterInfo` via `org.testng.ITestNGMethod#getFactoryMethodParamsInfo`. 

So if I move the `getParameters()` into `ITestNGMethod` then it would look as if the parameters are associated with all types of `ITestNGMethod`. That was why it was housed in `IParameterInfo`.

Currently TestNG does not have different interfaces for Configuration, factory and test methods. All of them are of type `ITestNGMethod`

Reply:
We are agree that some classes are missing in the current model.
But it could not be an arg to release a wrong model.
Here, the method can be removed from the interface and called with a cast or instanceof check. It will allow us to replace it by a better implementation later without the need to modify the public api.

Reply:
If it helps, you can keep it in the internal interface.

Reply:
The parameters are associated with the object that was instantiated (basically the values that were passed to the object's constructors). So i am not sure if it would make sense for us to the move them to the factory method level because a factory method can have parameters, which can take `n` values (if it were to be powered by a data provider for e.g.,). So I think the parameters represents the values that were used to create the object.